boa/src/syntax/parser/statement/mod.rs

```diff
@@ -289,6 +290,53 @@ where
             while cursor.next_if(Punctuator::Semicolon)?.is_some() {}
         }
 
+        // Handle any redeclarations
+        // https://tc39.es/ecma262/#sec-block-static-semantics-early-errors
+        {
+            let mut lexically_declared_names: HashSet<&str> = HashSet::new();
+            let mut var_declared_names: HashSet<&str> = HashSet::new();
+
+            // TODO: Use more helpful positions in errors when spans are added to Nodes
+            for item in &items {
+                match item {
+                    Node::LetDeclList(decl_list) | Node::ConstDeclList(decl_list) => {
+                        for decl in decl_list.as_ref() {
+                            // if name in VarDeclaredNames or can't be added to
+                            // LexicallyDeclaredNames, raise an error
+                            if var_declared_names.contains(decl.name())
+                                || !lexically_declared_names.insert(decl.name())
+                            {
+                                return Err(ParseError::lex(LexError::Syntax(
+                                    format!("Redeclaration of variable `{}`", decl.name()).into(),
+                                    match cursor.peek(0)? {
+                                        Some(token) => token.span().end(),
+                                        None => Position::new(1, 1),
+                                    },
+                                )));
+                            }
```

Comment:
Why is this a lexer error? The lexer correctly understood the tokens, right? It's just a normal parsing / Syntax error. Same happens for `VarDeclList`.

Reply:
What would be a better way to get a syntax error here? ParseError doesn't have it as an option.